### PR TITLE
Add `bless of moon` feature

### PR DESCRIPTION
**Pull request by Konoyoru435**
Sorry, I deleted my fork repo because I was too lazy to deal with the conflict that caused my PR to be automatically closed, and now I'm re-creating a PR.

In a nutshell: Item ID 1202 (Moon Card) is now available, adding support for the `bless of moon` feature.

Working list and main ideas:
- [x] Records the card activation time, duration, and activation status, and returns rewards based on these parameters;
    - Given before the player enters the world;
    - Activate the status after using the item and increase the duration.
- [x] Only one card can be used at a time;
- [x] If a player with an active moon card waits in the game until the next day, then the server should also send packets. (Very rough code)
    - [x] Completes a `Task` that is almost identical to the `Command` to facilitate registration of the Task.

Protos:
- `CardProductRewardNotify.proto`

PR Here: https://github.com/Grasscutters/Grasscutter-Protos/pull/10

Testing Work:
Testing...

### CODE REVIEW

File: src/main/java/emu/grasscutter/utils/DateHelper.java [hunk 1]
```diff
@@ -0,0 +1,16 @@
+package emu.grasscutter.utils;
+
+import java.util.Date;
+import java.util.Calendar;
+
+public final class DateHelper {
+    public static Date onlyYDM(Date now) {
```
Comment (WetABQ) [inline src/main/java/emu/grasscutter/utils/DateHelper.java]: what does this function mean

File: src/main/java/emu/grasscutter/task/tasks/MoonCard.java [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+import emu.grasscutter.game.GenshinPlayer;
+import emu.grasscutter.task.Task;
+import emu.grasscutter.task.TaskHandler;
+
+import java.util.List;
+
+import org.quartz.JobExecutionContext;
+import org.quartz.JobExecutionException;
+
+@Task(taskName = "MoonCard", taskCronExpression = "0 0 0 * * ?", triggerName = "MoonCardTrigger")
```
Comment (WetABQ) [inline src/main/java/emu/grasscutter/task/tasks/MoonCard.java:13]: could u add some comments on `taskCronExperssion`, it is a little bit abstract

File: src/main/java/emu/grasscutter/task/TaskMap.java [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+    }
+
+    public TaskHandler getHandler(String taskName) {
+        return this.tasks.get(taskName);
+    }
+
+    private void scan() {
+        Reflections reflector = Grasscutter.reflector;
+        Set<Class<?>> classes = reflector.getTypesAnnotatedWith(Task.class);
+        System.out.println("Found " + classes.size() + " tasks.");
```
Comment (WetABQ) [inline src/main/java/emu/grasscutter/task/TaskMap.java]: plz del the debug code

File: src/main/java/emu/grasscutter/task/Task.java [hunk 4]
```diff
@@ -0,0 +1,11 @@
+package emu.grasscutter.task;
+
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+
+@Retention(RetentionPolicy.RUNTIME)
+public @interface Task {
+    String taskName() default "NO_NAME";
+    String taskCronExpression() default "0 0 0 0 0 ?";
```
Comment (WetABQ) [inline src/main/java/emu/grasscutter/task/Task.java:28]: comments on here also

File: src/main/java/emu/grasscutter/utils/DateHelper.java (hunk 1 above)
Comment (Yazawazi) [reply src/main/java/emu/grasscutter/utils/DateHelper.java]: Keep the Date information as month, day and year only

File: src/main/java/emu/grasscutter/task/tasks/MoonCard.java (hunk 2 above)
Comment (Yazawazi) [reply src/main/java/emu/grasscutter/task/tasks/MoonCard.java:13]: https://en.wikipedia.org/wiki/Cron

File: src/main/java/emu/grasscutter/utils/DateHelper.java (hunk 1 above)
Comment (Konoyoru435) [reply src/main/java/emu/grasscutter/utils/DateHelper.java]: > what does this function mean

Keep the Date information as month, day and year only

File: src/main/java/emu/grasscutter/task/tasks/MoonCard.java (hunk 2 above)
Comment (Konoyoru435) [reply src/main/java/emu/grasscutter/task/tasks/MoonCard.java:13]: > could u add some comments on `taskCronExperssion`, it is a little bit abstract

https://en.wikipedia.org/wiki/Cron

File: src/main/java/emu/grasscutter/task/TaskMap.java (hunk 3 above)
Comment (Konoyoru435) [reply src/main/java/emu/grasscutter/task/TaskMap.java]: > plz del the debug code

OKAY!

**Review by KingRainbow44 — APPROVED** [review]

**Review by ghost — APPROVED** [review]

**Review by memetrollsXD — APPROVED** [review]

**Comment by Yazawazi** [comment]
okay, guys. The task was successfully registered, but surprisingly `player.isOnline()` is always false, what's the deal? Can anyone help me? 😢

**Comment by Yazawazi** [comment]
try to use `synchronized` now

**Pull request merged**